### PR TITLE
fixes #34; Remove number and string literal suffixes

### LINKED ISSUE #34: Remove number and string literal suffixes
These are a weird unnecessary special case for a Lisp that cause unnecessary complexity. Instead number should be like this:

- With a `u` suffix it is an unsigned number.
- Else if the number has a dot or an `E` it is a floating point number.
- Else it is an integer.

Typed integer literals then can be modelled simply as `(suf 234 "u32")`. This takes up more space but this should be irrelevant.

### PR DESCRIPTION
**Pull request by ringabout**
fixes #34

- [x] nif
- [x] nifreader
- [x] nifbuilder
- [x] ......
- [x] remove suffix for ints and floats from gear2, nifc and undone the APIs
- [x] documentaion

```nim
echo (12'u32, 12.3, 12e2, 7.5f32, "123", r"debility")
```
nif sample output
```
(.nif24)
(.vendor "Nifler")
(.dialect "nim-parsed")
@0,1,/home/wind/defs/Nim/nif/test.nim(stmts
 (cmd
 echo @5(tupleConstr
 @1 (suf 12 "u32")@9 12.3 @15 1200.0 @21 7.5 @29"123" @36(suf
 "debility" "R"
)
)
)
)
```
nifbuilder sample output
```
(.nif24)
(.vendor "tester")
(.dialect "niftest")

(stmts @4,5,mymodb.nim
 (call @1,3,mymod.nim foo.3.mymod 3423
  (suf 3423 "u32")
```
nifreader sample
```
(stmts
 (cmd
  Ident:echo (tupleConstr
   (suf
    IntLit:12 StringLit:u32) FloatLit:12.3 FloatLit:1200.0 StringLit:123 (suf
    StringLit:debility StringLit:R))))
```

### CODE REVIEW

File: src/lib/nifreader.nim [hunk 1]
```diff
@@ -316,6 +309,26 @@ proc next*(r: var Reader): Token =
         let repl = r.ksubs.getOrDefault(result.s)
         if repl[0] != UnknownToken:
           result.s = repl[1]
+      elif result.s == "suf":
```
Comment (Araq) [inline src/lib/nifreader.nim:312]: No. Remove. It's not a special thing for the reader.

File: src/nifler/emitter.nim [hunk 2]
```diff
@@ -126,6 +108,45 @@ proc addEmpty*(em: var Emitter; count = 1) =
...
+      inc lastPart, 3
+    else:
+      r.add c
+      inc l, 1
+      inc lastPart, 1
+    afterNewline = false
+  r.add "\""
+  em.emit r, lastPart
+
+proc addStrLit*(em: var Emitter; s, suffix: string) =
```
Comment (Araq) [inline src/nifler/emitter.nim:146]: No need for this API.

Comment (ringabout) [reply src/nifler/emitter.nim:146]: The old one provides this API
```nim
proc addStrLit*(em: var Emitter; s, suffix: string)
```
Shouldn't strings have suffix?

File: doc/nif-spec.md [hunk 3]
```diff
@@ -194,13 +194,15 @@ Grammar:
 
 ```
 Digit ::= [0-9]
-NumberSuffix ::= [a-z]+ [0-9a-z]*  # suffixes can only contain lowercase letters
+NumberSuffix ::= '"' [a-z]+ [0-9a-z]* '"' # suffixes can only contain lowercase letters
```
Comment (Araq) [inline doc/nif-spec.md:197]: NumberSuffix can only be `u` for "unsigned", nothing else exists, the `suf` tree is not part of the lexing step.

Comment (ringabout) [reply doc/nif-spec.md:197]: >  the suf tree is not part of the lexing step.

What does this mean? Do you mean the suf tree is handled by nifc/gear2/...?

Comment (Araq) [reply doc/nif-spec.md:197]: Yes.

Comment (ringabout) [reply doc/nif-spec.md]: eh, I cannot compile gear2

Comment (ringabout) [reply doc/nif-spec.md]: I have fixed gear2/gen2nif. `genToNim` seems to be incomplete

File: src/nifc/nifc_model.nim [hunk 4]
```diff
@@ -163,12 +164,23 @@ proc parse*(r: var Reader; dest: var PackedTree[NifcKind]; m: var Module; parent
     result = false
   of ParLe:
     let kind = whichNifcKeyword(t.s, Err)
-    var d = if kind == TypeC: addr(m.types) else: addr(dest)
-    copyInto(d[], kind, currentInfo):
-      while true:
-        let progress = parse(r, d[], m, currentInfo)
-        if not progress: break
-
+    if kind == SufC:
```
Comment (Araq) [inline src/nifc/nifc_model.nim:167]: No reason for this special logic here.

Comment (ringabout) [reply src/nifc/nifc_model.nim:167]: Okay, I can let codegen handle `SufC`

Comment (Araq) [reply src/nifc/nifc_model.nim:167]: Exactly.

**Comment by Araq** [comment]
Ok, taking over from here.

**Pull request merged**